bindings/rust/bench/src/s2n_tls.rs

```diff
@@ -0,0 +1,213 @@
+            }
+            Mode::Server => {
+                read_ptr = client_to_server_ptr;
+                write_ptr = server_to_client_ptr;
+                config = &self.server_config;
+                conn = &mut self.server_conn;
+            }
+        }
+
+        conn.set_blinding(Blinding::SelfService)?
+            .set_config(config.clone())?
+            .set_send_callback(Some(Self::send_cb))?
+            .set_receive_callback(Some(Self::recv_cb))?;
+        unsafe {
+            conn.set_send_context(write_ptr)?
+                .set_receive_context(read_ptr)?;
+        }
+
+        Ok(())
+    }
+
+    /// Handshake step for one connection
+    fn handshake_conn(&mut self, mode: Mode) {
+        let (conn, handshaked);
+        match mode {
+            Mode::Client => {
+                debug!("Client: ");
+                conn = &mut self.client_conn;
+                handshaked = &mut self.client_handshake_completed;
+            }
+            Mode::Server => {
+                debug!("Server: ");
+                conn = &mut self.server_conn;
+                handshaked = &mut self.server_handshake_completed;
+            }
+        }
+        if let Ready(res) = conn.poll_negotiate() {
+            res.unwrap();
+            *handshaked = true;
+        } else {
+            *handshaked = false;
+        }
+    }
+}
+
+impl TlsBenchHarness for SignalToNoise {
+    fn new() -> Self {
+        debug!("----- constructing new s2n-tls harness -----");
+        let mut new_harness = SignalToNoise {
+            client_to_server_buf: Box::pin(UnsafeCell::new(VecDeque::new())),
+            server_to_client_buf: Box::pin(UnsafeCell::new(VecDeque::new())),
+            client_config: Self::create_config(Mode::Client),
+            server_config: Self::create_config(Mode::Server),
+            client_conn: Connection::new_client(),
+            server_conn: Connection::new_server(),
+            client_handshake_completed: false,
+            server_handshake_completed: false,
+        };
+        new_harness.init_conn(Mode::Client).unwrap();
+        new_harness.init_conn(Mode::Server).unwrap();
+        new_harness
+    }
+
+    fn handshake(&mut self) -> Result<(), &str> {
+        if self.handshake_completed() {
+            return Err("Already completed handshake");
+        }
+        debug!("HANDSHAKE");
+        let mut round_trips = 2; // expect two round trips, second for server to see Finished message
+        while !self.handshake_completed() && round_trips > 0 {
+            self.handshake_conn(Mode::Client);
+            self.handshake_conn(Mode::Server);
+            round_trips -= 1;
+        }
+        debug!("HANDSHAKE DONE\n");
+        Ok(())
+    }
+
+    fn handshake_completed(&self) -> bool {
+        self.client_handshake_completed && self.server_handshake_completed
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn s2n_tls_create_object() {
```

Comment:
There's a lot of unwrapping that happens when you make a new object, so I wanted a separate test for whether making a new harness panics or not; in the future when things get more complex, I would probably have `new()` return a Result instead of panic.